File: spark/src/main/scala/cn/fancychuan/scala/apps/WordCountApp.scala
package cn.fancychuan.scala.apps

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

object WordCountApp {
  def main(args: Array[String]): Unit = {
    var args1: Array[String] = new Array[String](2)
    if (args.length == 0) {
      args1 = Array("hadoop/input/wordcount.txt", "spark/target/wordcount")
      println(args1)
    } else {
      args1 = args
    }
    val config: SparkConf = new SparkConf().setMaster("local[*]").setAppName("SparkWordCount")

    val sc = new SparkContext(config)
    println(sc)

    // 这里地方路径跟spark部署方式有关。比如配置的是yarn，那么会读取HDFS上的路径
    // 如果我们需要本地文件系统，那么要写成 file:///opt/module/xxxxx
    val lines: RDD[String] = sc.textFile(args1(0))
    val words: RDD[String] = lines.flatMap(_.split(" "))

    val wordToOne: RDD[(String, Int)] = words.map((word) => (word, 1))

    val wordToSum: RDD[(String, Int)] = wordToOne.reduceByKey(_ + _)

    val result: Array[(String, Int)] = wordToSum.collect()

    print(result)
    result.foreach(println)

    wordToSum.saveAsTextFile(args1(1))
  }
}

File: spark/spark2.4.4/src/main/java/cn/fancychuan/spark244/wordcount/WordCountApp.java
package cn.fancychuan.spark244.wordcount;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;

import java.util.Arrays;

public class WordCountApp {

    public static void main(String[] args) {
        if (args.length == 0) {
            args = new String [] {"spark/data/wordcount.txt", "spark/target/wordcount"};
        }
        SparkConf sparkConf = new SparkConf().setAppName("wordcount-java");
        JavaSparkContext sc = new JavaSparkContext(sparkConf);
        JavaRDD<String> wordsRDD = sc.textFile(args[0]).flatMap(line -> Arrays.asList(line.split(" ")).iterator());
        JavaPairRDD<String, Long> resultRDD = wordsRDD.mapToPair(word -> new Tuple2<>(word, 1L)).reduceByKey((x, y) -> x + y);
        System.out.println("=========== gogogo ============");
        System.out.println(resultRDD.collect());
        System.out.println("===============================");
        resultRDD.saveAsTextFile(args[1]);
    }
}
